source/common/http/http2/codec_impl.cc

```diff
@@ -900,51 +905,78 @@ void ConnectionImpl::sendPendingFrames() {
-    ENVOY_CONN_LOG(debug, "setting max concurrent streams to {}", connection_,
-                   http2_settings.max_concurrent_streams_);
-  }
-
-  if (http2_settings.initial_stream_window_size_ != NGHTTP2_INITIAL_WINDOW_SIZE) {
-    iv.push_back(
-        {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_settings.initial_stream_window_size_});
-    ENVOY_CONN_LOG(debug, "setting stream-level initial window size to {}", connection_,
-                   http2_settings.initial_stream_window_size_);
-  }
+void ConnectionImpl::sendSettings(
+    const envoy::config::core::v3::Http2ProtocolOptions& http2_options, bool disable_push) {
+  std::unordered_set<nghttp2_settings_entry, SettingsEntryHash, SettingsEntryEquals> settings;
 
+  // Universally disable receiving push promise frames as we don't currently support
+  // them. nghttp2 will fail the connection if the other side still sends them.
+  // TODO(mattklein123): Remove this when we correctly proxy push promise.
+  // NOTE: This is a special case with respect to custom parameter overrides in that server push is
+  // not supported and therefore not end user configurable.
   if (disable_push) {
-    // Universally disable receiving push promise frames as we don't currently support them. nghttp2
-    // will fail the connection if the other side still sends them.
-    // TODO(mattklein123): Remove this when we correctly proxy push promise.
-    iv.push_back({NGHTTP2_SETTINGS_ENABLE_PUSH, 0});
+    settings.insert({static_cast<int32_t>(NGHTTP2_SETTINGS_ENABLE_PUSH), disable_push ? 0U : 1U});
+  }
+
+  for (const auto it : http2_options.custom_settings_parameters()) {
+    ASSERT(it.identifier().value() <= std::numeric_limits<uint16_t>::max());
+    if (it.identifier().value() == NGHTTP2_SETTINGS_ENABLE_PUSH && it.value().value() == 1) {
+      // For simplicity, warn when server push is configured enabled on either client or server
+      // connections. It _is_ actually enabled on downstream connections although Envoy itself will
+      // never send push_promise frames, but the nuance could be confusing to end users.
+      ENVOY_CONN_LOG(
+          warn,
+          "server push is not supported by Envoy and can not be enabled via a SETTINGS parameter.",
+          connection_);
+      continue;
+    }
+    auto result =
+        settings.insert({static_cast<int32_t>(it.identifier().value()), it.value().value()});
+    if (!result.second) {
+      ENVOY_CONN_LOG(debug, "duplicate user defined settings parameter with id {:#x}, value {}",
+                     connection_, it.identifier().value(), it.value().value());
+      continue;
+    }
+    ENVOY_CONN_LOG(debug, "adding user defined settings parameter with id {:#x} to {}", connection_,
+                   it.identifier().value(), it.value().value());
+  }
+
+  struct SettingsDescriptor {
+    nghttp2_settings_entry entry;
+    uint32_t default_value;
+  };
+  auto descriptors = std::vector<SettingsDescriptor>{
```

Comment:
perf nit: given that the number of entries here is going to be tiny, I would maybe use InlinedVector? I'm wondering if you can also remove the map as I'm guessing it doesn't offer much value but up to you.

Reply:
Done.